server/src/main/java/org/apache/uniffle/server/ShuffleServerGrpcService.java

```diff
@@ -157,6 +157,36 @@ public void registerShuffle(
     int shuffleId = req.getShuffleId();
     String remoteStoragePath = req.getRemoteStorage().getPath();
     String user = req.getUser();
+    int stageAttemptNumber = req.getStageAttemptNumber();
+
+    if (stageAttemptNumber > 0) {
+      ShuffleTaskInfo taskInfo = shuffleServer.getShuffleTaskManager().getShuffleTaskInfo(appId);
+      int attemptNumber = taskInfo.getLatestStageAttemptNumber(shuffleId);
+      if (stageAttemptNumber > attemptNumber) {
+        taskInfo.refreshLatestStageAttemptNumber(shuffleId, stageAttemptNumber);
+        try {
+          long start = System.currentTimeMillis();
+          shuffleServer.getShuffleTaskManager().removeShuffleDataSync(appId, shuffleId);
```

Comment:
If this costs too much time, maybe we could use the rename , but the purge should be guaranteed.

Maybe the shuffleId folder could be rename to negative number, that means it's waiting for purge async.

Reply:
> If this costs too much time, maybe we could use the rename , but the purge should be guaranteed.
> 
> Maybe the shuffleId folder could be rename to negative number, that means it's waiting for purge async.

This function is more modified, I plan to get a separate PR to complete.

Reply:
Good to know this..